library/src/main/kotlin/com/connectrpc/impl/ProtocolClient.kt

```diff
@@ -197,32 +200,51 @@ class ProtocolClient(
             // Pass through the interceptor chain.
             when (val streamResult = streamFunc.streamResultFunction(initialResult)) {
                 is StreamResult.Headers -> {
-                    // Not currently used except for interceptors.
+                    // If this is incorrectly called 2x, only the first result is used.
+                    // Subsequent calls to complete will be ignored.
+                    responseHeaders.complete(streamResult.headers)
                 }
 
                 is StreamResult.Message -> {
+                    // Just in case protocol impl failed to provide StreamResult.Headers,
+                    // treat headers as empty. This is a no-op if we did correctly receive
+                    // them already.
+                    responseHeaders.complete(emptyMap())
                     try {
                         val message = responseCodec.deserialize(
                             streamResult.message,
                         )
                         channel.send(message)
                     } catch (e: Throwable) {
+                        // TODO: setting isComplete, responseTrailers, and RPC status
+                        //       here seems wrong. What would prevent the call to
+                        //       channel.send such that we don't bother getting the
+                        //       actual result/trailers from the server?
```

Comment:
I think in this case it is more likely that we couldn't deserialize the response, not that we couldn't send on the channel. The only way the send would fail would be if we received the Complete stream result before the message.

Reply:
> The only way the send would fail would be if we received the Complete stream result before the message.

After I wrote this, I think I saw a comment somewhere suggesting that this could happen if the message can't be serialized. I'm pretty sure that can happen, among other scenarios, when JSON is used and the message is or contains a `google.protobuf.Any` that contains a message whose schema is not known to this program. (I'm not even sure how that is configured since the Java runtime has no global registry -- I think you may need a message registry configured with the serializers/deserializers, and that's what it consults when it needs to transcode an `Any`). If we already received the Complete stream, I think we'd get a different error -- like an error about the stream already being closed.